### PR TITLE
Bundling/streamlining requests for indices page.

### LINKED ISSUE #1959: Index Summary needs more information to distinguish "normal"/closed/reopened indices
### Problem description

The indices page needs several improvements.

1. A closed index shows `Time range of index is unknown, because index range is not available. Please recalculate index ranges manually.` in the overview. No other sign that it's closed until you open the index details. There should be an icon that indicates that the index is closed. Also the text might be improved to not look like an error.
cc6o6je9naiaaaaasuvork5cyii
2. Use proper labels for closed/re-opened indices instead of the "retweet" icon.

### Steps to reproduce the problem

1. Navigate to "System/Indices"
2. Close an index
3. Wait for auto reload or reload manually to see the message.

### Environment

* Graylog Version: Graylog 2.0.0-alpha.6-SNAPSHOT (5bd54d0)

### LINKED ISSUE #1960: Indices overview page gets slow when there are many indices
### Problem description

The Indices details are currently fetched over a single REST bulk endpoint providing information about all indices. Therefore all data needs to be refetched every time the details for a single index are updated. This slows down data fetching from the server and therefore page updates.

### Steps to reproduce the problem

1. Create a lot of indices (~three figure count)
2. Open "System / Indices" page
3. See big responses coming from the server slowing down processing

### Environment

* Graylog Version: 2.0.0-alpha.5
* Elasticsearch Version:
* MongoDB Version:
* Operating System:
* Browser version:

### PR DESCRIPTION
**Pull request by dennisoelkers**
* Add endpoints for indices overview and multiple index details.
* Add new store/actions which make use of new index/indices overview.
* Allows components to (un-)/register for index details.
* Use new index subscriptions and index/indices overview in components.

Fixes #1959, #1960

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog2/rest/resources/system/indexer/IndexerOverviewResource.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        this.deflectorResource = deflectorResource;
+        this.indexerClusterResource = indexerClusterResource;
+        this.indexRangesResource = indexRangesResource;
+        this.countResource = countResource;
+        this.indices = indices;
+    }
+
+    @GET
+    @Timed
+    @ApiOperation(value = "Get overview of current indexing state, including deflector config, cluster state, index rnages & message counts.")
```
Comment (edmundoa) [inline graylog2-server/src/main/java/org/graylog2/rest/resources/system/indexer/IndexerOverviewResource.java]: Ranges is misspelled :)

File: graylog2-web-interface/src/stores/indices/IndicesStore.js [hunk 2]
```diff
@@ -29,6 +30,27 @@ const IndicesStore = Reflux.createStore({
...
+      } else {
+        Object.keys(response).forEach((indexName) => {
+          this.indices[indexName] = response[indexName];
+        });
+      }
+      this.trigger({ indices: this.indices, closedIndices: this.closedIndices });
+      return { indices: this.indices, closedIndices: this.closedIndices };
+    });
+
+    IndicesActions.list.promise(promise);
```
Comment (edmundoa) [inline graylog2-web-interface/src/stores/indices/IndicesStore.js]: Shouldn't this method use the `multiple` action to store the promise?

File: graylog2-web-interface/src/stores/indexers/IndexerOverviewStore.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import ActionsProvider from 'injection/ActionsProvider';
+const IndexerOverviewActions = ActionsProvider.getActions('IndexerOverview');
+
+const IndexerOverviewStore = Reflux.createStore({
+  listenables: [IndexerOverviewActions],
+  list() {
+    const url = URLUtils.qualifyUrl(ApiRoutes.IndexerOverviewApiResource.list().url);
+    const promise = fetch('GET', url);
+    promise.then((response) => {
+      this.trigger(response);
```
Comment (edmundoa) [inline graylog2-web-interface/src/stores/indexers/IndexerOverviewStore.js]: I don't really like this, as the response is returning several keys that will go into a component's state (at least if it's connecting to the store), and there's no obvious way of knowing which keys they are unless you look into the response. Can we make the keys explicit here?

File: graylog2-web-interface/src/components/indices/IndicesOverview.jsx [hunk 4]
```diff
@@ -8,21 +8,25 @@ const IndicesOverview = React.createClass({
   propTypes: {
     closedIndices: React.PropTypes.array.isRequired,
     deflector: React.PropTypes.object.isRequired,
-    indexRanges: React.PropTypes.array.isRequired,
+    indexDetails: React.PropTypes.object.isRequired,
     indices: React.PropTypes.object.isRequired,
   },
   _isDeflector(index) {
```
Comment (edmundoa) [inline graylog2-web-interface/src/components/indices/IndicesOverview.jsx]: This function is unused now, we can remove it.

File: graylog2-web-interface/src/components/indices/IndexSummary.jsx [hunk 5]
```diff
@@ -9,28 +11,37 @@ const IndexSummary = React.createClass({
...
-      return 'Contains messages up to ' + new DateTime().toRelativeString();
+      return `Contains messages up to ${new DateTime().toRelativeString()}`;
     }
 
-    if (this.props.index.all_shards) {
-      const count = this.props.index.all_shards.documents.count;
-      const deleted = this.props.index.all_shards.documents.deleted;
+    const sizes = this.props.index.size;
+    if (sizes) {
+      const count = sizes.count;
```
Comment (edmundoa) [inline graylog2-web-interface/src/components/indices/IndexSummary.jsx]: Is this calculation right? It looks like `count` is not part of `IndexSizeSummary`.

File: graylog2-web-interface/src/components/indices/IndexSummary.jsx [hunk 6]
```diff
@@ -9,28 +11,37 @@ const IndexSummary = React.createClass({
...
-  },
-  _reopenedBadged(index) {
-    if (index.reopened) {
-      return <i className="fa fa-retweet" title="Reopened index"/>;
+
+    if (index.is_closed) {
+      labels.push(<Label key={`${this.props.name}-closed-label`} bsStyle="warning">closed</Label>);
+    }
+
+    if (index.is_reopened) {
```
Comment (edmundoa) [inline graylog2-web-interface/src/components/indices/IndexSummary.jsx]: When a reopened index is closed again, we display two labels for it, even though we should only display closed in that case.

File: graylog2-web-interface/src/components/indices/IndexSummary.jsx [hunk 7]
```diff
@@ -61,13 +72,10 @@ const IndexSummary = React.createClass({
...
-          {index.name}{' '}
-
-          {this._reopenedBadged(index)}{' '}
+          {this.props.name}{' '}
 
           <small>
+            {this._formatLabels(index)}{' '}
             {this._formatIndexRange(index)}{' '}
 
             <IndexSizeSummary index={index} />
```
Comment (edmundoa) [inline graylog2-web-interface/src/components/indices/IndexSummary.jsx]: `IndexSizeSummary` is not updated to the new data structure coming from the API, and it's never rendered.

File: graylog2-web-interface/src/components/indices/IndexSummary.jsx (hunk 6 above)
Comment (dennisoelkers) [inline graylog2-web-interface/src/components/indices/IndexSummary.jsx]: No, this is not the case, as the backend does not list a reopened index as closed.

File: graylog2-server/src/main/java/org/graylog2/rest/resources/system/indexer/IndexerOverviewResource.java (hunk 1 above)
Comment (dennisoelkers) [inline graylog2-server/src/main/java/org/graylog2/rest/resources/system/indexer/IndexerOverviewResource.java]: Oh, it's not spelled rnages? :)

File: graylog2-web-interface/src/stores/indices/IndicesStore.js (hunk 2 above)
Comment (dennisoelkers) [inline graylog2-web-interface/src/stores/indices/IndicesStore.js]: :+1:

File: graylog2-web-interface/src/components/indices/IndexSummary.jsx (hunk 7 above)
Comment (dennisoelkers) [inline graylog2-web-interface/src/components/indices/IndexSummary.jsx]: :+1: Fixed in 522f41275112e304d3e46833375502fb50f4edfa

File: graylog2-web-interface/src/components/indices/IndexSummary.jsx (hunk 6 above)
Comment (edmundoa) [inline graylog2-web-interface/src/components/indices/IndexSummary.jsx]: Probably it shouldn't, but it does:
```
"graylog2_0": {
  "size":null,
  "range":null,
  "is_closed":true,
  "is_reopened":true,
  "is_deflector":false
}
```

screen shot 2016-04-04 at 14 39 10

File: graylog2-web-interface/src/stores/indexers/IndexerOverviewStore.js (hunk 3 above)
Comment (dennisoelkers) [inline graylog2-web-interface/src/stores/indexers/IndexerOverviewStore.js]: Agreed!

File: graylog2-web-interface/src/components/indices/IndexSummary.jsx (hunk 6 above)
Comment (dennisoelkers) [inline graylog2-web-interface/src/components/indices/IndexSummary.jsx]: True. Filed an issue (#2019) and worked around this in the IndexerOverviewResource.

**Comment by edmundoa** [comment]
LGTM :+1:

**Pull request merged**